### PR TITLE
Menhir Incremental parser

### PR DESCRIPTION
**Pull request by apoorvingle**
- Uses Menhir's Incremental API instead of monolith.

- Better error reporting for syntax errors, mentions line number and column along with file name and custom generated error messages.
- Generates useful error messages for: 
   - missing semicolon
   - colon between types
   - missing types
   - mismatched parenthesis
   - missing `return` clause on node/arguments to `pre` 
- `src/lustre/Readme.md` specifies how to add new error messages.

-  [Benchmarking log](https://github.com/kind2-mc/kind2/files/4832398/hf-kind2-bm.log): No impact

### CODE REVIEW

File: src/lustre/lustreInput.ml [hunk 1]
```diff
@@ -17,123 +17,136 @@
...
+    match LPMI.stack env with
+    | lazy Nil -> "Syntax Error"
+    | lazy (Cons (LPMI.Element (state, _, _, _), _)) ->
+       let pstate = LPMI.number state in
+        try (LPE.message pstate) with
+        | Not_found -> "Syntax Error (Parser State: "^ (string_of_int pstate) ^ ")" 
+                       ^ "Please report this issue with a minimum working example."
+
+(* Raises the [Parser_error] exception with appropriate position and error message *)
+let fail env lexbuf =
```
Comment (apoorvingle) [inline src/lustre/lustreInput.ml:58]: This should use `LC.fail_at_position` instead

File: src/lustre/lustreInput.ml [hunk 2]
```diff
@@ -17,123 +17,136 @@
...
+     parse lexbuf chkpnt
+  | LPMI.Shifting _
+  | LPMI.AboutToReduce _ ->
+     let chkpnt = LPMI.resume chkpnt in
+     parse lexbuf chkpnt
+  | LPMI.HandlingError env ->
+     fail env lexbuf
+  | LPMI.Accepted v -> success v
+  | LPMI.Rejected ->
+     raise (LA.Parser_error {position=None; err="Parser Error: Parser rejected the input."})
```
Comment (apoorvingle) [inline src/lustre/lustreInput.ml:79]: use `LC.fail_at_position`

File: .gitignore [hunk 3]
```diff
@@ -37,3 +37,7 @@ doc/user_documentation.pdf
 # GitHub deploy key
 travis_key
 travis_key.pub
+
+# merlin files
+.merlin
+./**/.merlin
```
Comment (daniel-larraz) [inline .gitignore:43]: Duplicate rules

File: .merlin [hunk 4]
```diff
@@ -0,0 +1,10 @@
+S src
+S src/**
+S src/lustre
+
+B _build
+B _build/**
+B _build/src
+B _build/src/lustre
+
+PKG menhir menhirLib yojson
```
Comment (daniel-larraz) [inline .merlin:10]: Remove file

File: src/.merlin [hunk 5]
```diff
@@ -0,0 +1,8 @@
+B _build
```
Comment (daniel-larraz) [inline src/.merlin:1]: Remove file

File: src/kind2.ml [hunk 6]
```diff
@@ -96,22 +96,21 @@ let setup : unit -> any_input = fun () ->
...
-    some stuff. *)
-    KEvent.terminate_log () ;
-    exit ExitCodes.error
+  | LustreAst.Parser_error  ->
+     (* terminate log after printing file name and exit as the error is already on the log file *)
+     KEvent.log L_fatal "While parsing file %s" (Flags.input_file ()); 
+     KEvent.terminate_log () ; exit ExitCodes.error
+  | LustreLexer.Lexer_error err ->
+     KEvent.log L_fatal "Lexing error: %s" err;
+     KEvent.terminate_log () ; exit ExitCodes.error
```
Comment (apoorvingle) [inline src/kind2.ml:105]: missing file name.

Comment (apoorvingle) [reply src/kind2.ml:105]: check why Lexing error was not printed.

File: src/lustre/README.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+The following steps are used to update the `LustreParserErrors.message` function
+
+1. Edit the error message in `src/lustre/lustreParser.message` file
+2. Generate the function using the following command from (project root directory)
+```
+$ menhir --compile-errors src/lustre/lustreParser.messages src/lustre/lustreParser.mly > src/lustre/lustreParserErrors.ml
+```
+3. Compile against the new `LustreParserErrors.ml` file
+```
+$ make kind2
```
Comment (apoorvingle) [inline src/lustre/README.md:15]: now uses dune,  so change to `make build`

File: src/lustre/lustreParserErrors.mli [hunk 8]
```diff
@@ -0,0 +1,22 @@
+(* This file is part of the Kind 2 model checker.
+
+   Copyright (c) 2015 by the Board of Trustees of the University of Iowa
```
Comment (apoorvingle) [inline src/lustre/lustreParserErrors.mli:3]: year of copy right

File: src/kind2.ml (hunk 6 above)
Comment (apoorvingle) [reply src/kind2.ml:105]: New error message now has file name
```
<Error> While parsing file ./tests/lustre/syntax-errors/array-syntax.lus
 Lexing error: Unrecognized token % (0x25) found at line 4, column 5
```

File: .gitignore (hunk 3 above)
Comment (apoorvingle) [reply .gitignore]: done

File: .merlin (hunk 4 above)
Comment (apoorvingle) [reply .merlin]: done

File: src/.merlin (hunk 5 above)
Comment (apoorvingle) [reply src/.merlin]: done

File: src/lustre/lustreParserErrors.mli (hunk 8 above)
Comment (apoorvingle) [reply src/lustre/lustreParserErrors.mli]: done

File: src/kind2.ml [hunk 9]
```diff
@@ -96,22 +96,21 @@ let setup : unit -> any_input = fun () ->
...
       exit ExitCodes.error
   with (* Could not create input system. *)
-  | LustreAst.Parser_error ->
-    (* Don't do anything for parser error, they should already have printed
-    some stuff. *)
-    KEvent.terminate_log () ;
-    exit ExitCodes.error
+  | LustreAst.Parser_error  ->
+     (* terminate log after printing file name and exit as the error is already on the log file *)
+     KEvent.log L_fatal "While parsing file %s" (Flags.input_file ());
```
Comment (daniel-larraz) [inline src/kind2.ml:101]: You should check whether the filename is empty (it is reading from standard input), and adjust the message accordingly. Same comment for lexer error.

File: .gitignore [hunk 10]
```diff
@@ -13,6 +13,7 @@ _build
 
 # Generated merlin file.
 .merlin
+./**/.merlin
```
Comment (daniel-larraz) [inline .gitignore:16]: Pattern `.merlin` should already ignore all files (at any depth).

File: src/kind2.ml (hunk 9 above)
Comment (apoorvingle) [reply src/kind2.ml:101]: makes sense. fixed it now.
```
cat tests/lustre/syntax-errors/array-syntax.lus |  ./bin/kind2 -vv --only_parse true
 kind2 cf9613e-dirty

 Parsing standard input.

<Error> While parsing standard input
 Lexing error: Unrecognized token % (0x25) found at line 4, column 5
```
and while parsing from a file:

```
./bin/kind2 -vv --only_parse true tests/lustre/syntax-errors/array-syntax.lus
 kind2 cf9613e-dirty

 Parsing input file 'tests/lustre/syntax-errors/array-syntax.lus'.

<Error> While parsing input file 'tests/lustre/syntax-errors/array-syntax.lus'
 Lexing error: Unrecognized token % (0x25) found at line 4, column 5
```

File: .gitignore (hunk 10 above)
Comment (apoorvingle) [reply .gitignore]: done

**Pull request merged**